File: src/test/kotlin/ru/serdtsev/homemoney/moneyoper/MoneyOperServiceTest.java
package ru.serdtsev.homemoney.moneyoper;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.springframework.core.convert.ConversionService;
import ru.serdtsev.homemoney.account.dao.AccountDao;
import ru.serdtsev.homemoney.account.dao.BalanceDao;
import ru.serdtsev.homemoney.account.model.Account;
import ru.serdtsev.homemoney.account.model.AccountType;
import ru.serdtsev.homemoney.account.model.Balance;
import ru.serdtsev.homemoney.balancesheet.BalanceSheetDao;
import ru.serdtsev.homemoney.balancesheet.model.BalanceSheet;
import ru.serdtsev.homemoney.moneyoper.dao.MoneyOperDao;
import ru.serdtsev.homemoney.moneyoper.dao.RecurrenceOperDao;
import ru.serdtsev.homemoney.moneyoper.dao.TagDao;
import ru.serdtsev.homemoney.moneyoper.dto.MoneyOperDto;
import ru.serdtsev.homemoney.moneyoper.model.*;
import ru.serdtsev.homemoney.moneyoper.service.MoneyOperService;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.UUID;

import static java.util.Arrays.asList;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

class MoneyOperServiceTest {
  private MoneyOperService service;
  private MoneyOperDao moneyOperDao;
  private BalanceSheet balanceSheet;
  private AccountDao accountDao;
  private ConversionService conversionService;

  @BeforeEach
  void setUp() {
    var balanceSheetDao = mock(BalanceSheetDao.class);
    balanceSheet = new BalanceSheet();
    doReturn(balanceSheet).when(balanceSheetDao).findByIdOrNull(any());

    moneyOperDao = mock(MoneyOperDao.class);
    accountDao = mock(AccountDao.class);
    var balanceDao = mock(BalanceDao.class);
    conversionService = mock(ConversionService.class);
    service = new MoneyOperService(balanceSheetDao, moneyOperDao, mock(RecurrenceOperDao.class),
        accountDao, balanceDao, mock(TagDao.class));
  }

  @Test
  @Disabled
  void getTagsSuggest() {
    Tag car = newTag("car");
    Tag food = newTag("food");
    Tag clothes = newTag("clothes");
    List<Tag> tagsA = asList(car, food, clothes);
    List<Tag> tagsB = asList(food, clothes);
    List<Tag> tagsC = asList(clothes);
    List<MoneyOper> opers = asList(newMoneyOperWithTags(tagsA), newMoneyOperWithTags(tagsB), newMoneyOperWithTags(tagsC));
    when(moneyOperDao.findByBalanceSheetAndStatusAndPerformedGreaterThan(any(), any(), any()))
        .thenReturn(opers);
    Account account = new Account(UUID.randomUUID(), balanceSheet, AccountType.debit, "Some account name", LocalDate.now(), false);
    when(accountDao.findNameById(any())).thenReturn(account.getName());

    MoneyOper moneyOper = newMoneyOperWithTags(new ArrayList<>());
    MoneyOperDto moneyOperDto = conversionService.convert(moneyOper, MoneyOperDto.class);
    Collection<Tag> checkTags = service.getSuggestTags(moneyOper.getItems().get(0).getBalance().getId(),
            MoneyOperType.expense.name(), null, moneyOperDto.getTags());

    List<Tag> expectedTags = asList(clothes, food, car);
    Assertions.assertIterableEquals(expectedTags, checkTags);
  }

  private MoneyOper newMoneyOperWithTags(List<Tag> tags) {
    MoneyOper moneyOper = new MoneyOper(UUID.randomUUID(), balanceSheet, List.of(), MoneyOperStatus.doneNew, LocalDate.now(), 0,
        tags, null, Period.month);
    var created = LocalDate.now();
    Balance balance1 = new Balance(balanceSheet, AccountType.debit, "Some account name", BigDecimal.ZERO);
    Balance balance2 = new Balance(balanceSheet, AccountType.debit, "Some account name", BigDecimal.ZERO);
    moneyOper.addItem(balance1, BigDecimal.valueOf(1).negate(), LocalDate.now(), 0, UUID.randomUUID());
    moneyOper.addItem(balance2, BigDecimal.ONE, LocalDate.now(), 1, UUID.randomUUID());
    return moneyOper;
  }

  private Tag newTag(String name) {
    return new Tag(balanceSheet, name);
  }

}

File: src/test/kotlin/ru/serdtsev/homemoney/port/service/MoneyOperServiceTest.kt
package ru.serdtsev.homemoney.port.service

import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import org.junit.jupiter.api.Assertions.assertIterableEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import ru.serdtsev.homemoney.domain.model.account.Account
import ru.serdtsev.homemoney.domain.model.account.AccountType
import ru.serdtsev.homemoney.domain.model.account.Balance
import ru.serdtsev.homemoney.domain.model.balancesheet.BalanceSheet
import ru.serdtsev.homemoney.domain.model.moneyoper.*
import ru.serdtsev.homemoney.domain.repository.BalanceRepository
import ru.serdtsev.homemoney.domain.repository.MoneyOperRepository
import ru.serdtsev.homemoney.domain.repository.TagRepository
import ru.serdtsev.homemoney.infra.dao.AccountDao
import ru.serdtsev.homemoney.infra.dao.BalanceSheetDao
import java.math.BigDecimal
import java.time.LocalDate
import java.util.*

internal class MoneyOperServiceTest {
    private val balanceSheetDao: BalanceSheetDao = mock {  }
    private val tagRepository: TagRepository = mock {  }
    private val moneyOperRepository: MoneyOperRepository = mock {  }
    private val balanceSheet = BalanceSheet()
    private val accountDao: AccountDao = mock {  }
    private val service = MoneyOperService(moneyOperRepository, mock {  }, tagRepository)

    @BeforeEach
    fun setUp() {
        whenever(balanceSheetDao.findById(any())).thenReturn(balanceSheet)
    }

    @Test
    fun getTagsSuggest() {
        val car = newTag("car")
        val food = newTag("food")
        val clothes = newTag("clothes")
        val tagsA = listOf(car, food, clothes)
        val tagsB = listOf(food, clothes)
        val tagsC = listOf(clothes)

        val opers = listOf(newMoneyOperWithTags(tagsA), newMoneyOperWithTags(tagsB), newMoneyOperWithTags(tagsC))
        whenever(moneyOperRepository.findByBalanceSheetAndStatusAndPerformedGreaterThan(any(), any()))
            .thenReturn(opers)

        whenever(tagRepository.findByBalanceSheetOrderByName())
            .thenReturn(tagsA)

        val account = Account(UUID.randomUUID(), AccountType.debit,"Some account name",
            LocalDate.now(),false)
        whenever(accountDao.findNameById(any())).thenReturn(account.name)

        val moneyOper = newMoneyOperWithTags(ArrayList())
        val tags = moneyOper.tags.map { it.name }
        val actual = service.getSuggestTags(MoneyOperType.expense.name, null, tags)
        val expectedTags = listOf(car, food, clothes)
        assertIterableEquals(expectedTags, actual)
    }

    private fun newMoneyOperWithTags(tags: List<Tag>): MoneyOper {
        val moneyOper = MoneyOper(UUID.randomUUID(), mutableListOf(), MoneyOperStatus.doneNew,
            LocalDate.now(), 0, tags, null, Period.month)
        val balance1 = Balance(AccountType.debit, "Some account name", BigDecimal.ZERO)
        val balance2 = Balance(AccountType.debit, "Some account name", BigDecimal.ZERO)
        moneyOper.addItem(balance1, BigDecimal.valueOf(1).negate(), LocalDate.now(), 0, UUID.randomUUID())
        moneyOper.addItem(balance2, BigDecimal.ONE, LocalDate.now(), 1, UUID.randomUUID())
        return moneyOper
    }

    private fun newTag(name: String): Tag {
        return Tag(name).apply {
            this.isCategory = true
            this.categoryType = CategoryType.expense
        }
    }
}